### DISCUSSION TITLE
can't get trashed users to show up in select2_from_ajax field

### DISCUSSION THREAD

**Discussion by prescriptionlifeline** [Laravel-Backpack/community-forum#1139] [Q&A (Help)] [answered]
In my User CRUD I have this:

```php
        CRUD::field([
            'name' => 'spouse',
            'label' => 'Spouse',
            'type' => 'select2_from_ajax',
            'entity' => 'spouse',
            'attribute' => 'lastNameFirstName',
            'data_source' => backpack_url('user/fetch/spouse'),
            'method' => 'POST',
        ]);
```
Here' my FetchOperation:

```
    protected function fetchSpouse()
    {
        return $this->fetch([
            'model' => User::class, // required
            'searchable_attributes' => ['lastNameFirstName'],
            'paginate' => 10, // items to show per page
            'searchOperator' => 'LIKE',
            'query' => function($model) {
                return $model->withTrashed();
            } // to filter the results that are returned
        ]);
    }
```
And here's my spouse method in my User model:

```php
    public function spouse()
    {
        return $this->belongsTo(User::class, 'spouseID')->withTrashed();
    }
```
With all of this I can _add_ trashed spouses without issue BUT I just can't see trashed spouse when viewing non-trashed users.

If I do `User::findOrFail(...)->spouse` in `php artisan tinker` that returns the spouse even if the spouse isn't trashed. I just can't get the `select2_from_ajax` to behave the same way as `php artisan tinker`.

Any ideas?

**Comment by jcastroa87** [comment]
Based on the information provided in the knowledge sources, I can offer some insights and suggestions to address your issue with viewing trashed spouses for non-trashed users in the select2_from_ajax field.

1. Relationship Definition:
Your current relationship definition looks correct. You're using `belongsTo` and including `withTrashed()`, which should allow you to retrieve trashed spouses [Github Issue #5630](https://github.com/Laravel-Backpack/CRUD/issues/5630).

2. FetchOperation:
In your `fetchSpouse` method, you're using `withTrashed()` which is good. However, you might want to consider modifying your query to explicitly include trashed spouses. Here's a suggestion:

```php
protected function fetchSpouse(Request $request)
{
    return $this->fetch([
        'model' => User::class,
        'searchable_attributes' => ['lastNameFirstName'],
        'paginate' => 10,
        'searchOperator' => 'LIKE',
        'query' => function($model) {
            return $model->withTrashed()->where(function($query) {
                $query->whereNull('deleted_at')
                      ->orWhereNotNull('deleted_at');
            });
        }
    ]);
}
```

This modification ensures that both trashed and non-trashed spouses are included in the results.

3. Excluding Current User:
If you want to exclude the current user from the spouse options, you can modify your `fetchSpouse` method as shown in the [Community Forum discussion #1131](https://github.com/Laravel-Backpack/community-forum/discussions/1131):

```php
protected function fetchSpouse(Request $request)
{
    $exclude = $request->input('exclude');
    return $this->fetch([
        'model' => User::class,
        'searchable_attributes' => ['lastNameFirstName'],
        'paginate' => 10,
        'searchOperator' => 'LIKE',
        'query' => function($model) use ($exclude) {
            return $exclude ?
                $model->withTrashed()->where('id', '!=' , $exclude) :
                $model->withTrashed();
        }
    ]);
}
```

Then, update your field definition to include the `exclude` parameter:

```php
CRUD::field([
    'name' => 'spouse',
    'label' => 'Spouse',
    'type' => 'select2_from_ajax',
    'entity' => 'spouse',
    'attribute' => 'lastNameFirstName',
    'data_source' => backpack_url('user/fetch/spouse?exclude=' . $this->crud->getCurrentEntry()->id),
    'method' => 'POST',
]);
```

4. Debugging:
If you're still having issues, you might want to add some debugging to your `fetchSpouse` method to ensure it's being called and returning the expected results. You can use Laravel's `Log` facade to write debug information to your application's log file.

If these suggestions don't resolve your issue, it might be worth checking if there are any JavaScript errors in the browser console when the select2_from_ajax field is loaded or used, as suggested in the [Github Issue #4907](https://github.com/Laravel-Backpack/CRUD/issues/4907).

**Reply by prescriptionlifeline** [reply]
> FetchOperation:
> In your `fetchSpouse` method, you're using `withTrashed()` which is good. However, you might want to consider modifying your query to explicitly include trashed spouses. Here's a suggestion:

That didn't help:

```php
    protected function fetchSpouse(Request $request)
    {
        $exclude = $request->input('exclude');
        return $this->fetch([
            'model' => User::class, // required
            'searchable_attributes' => ['name'],
            'paginate' => 10, // items to show per page
            'searchOperator' => 'LIKE',
            'query' => function($model) {
                return $model->withTrashed()->where(function($query) {
                    $query->whereNull('deleted_at')
                          ->orWhereNotNull('deleted_at');
                });
            }
        ]);
    }
```

> If these suggestions don't resolve your issue, it might be worth checking if there are any JavaScript errors in the browser console when the select2_from_ajax field is loaded or used, as suggested in the https://github.com/Laravel-Backpack/CRUD/issues/4907.
There are no errors there.

There are no errors in the JS console or much of anything really.

**Reply by jcastroa87** [reply, marked as answer]
This was answer [here](https://github.com/Laravel-Backpack/community-forum/issues/1252)

Cheers.
